### ISSUE TITLE
No timeout or connection error (if internet is off/ unavailable) while downloading using Amplify.Storage.downloadFile()

### ISSUE THREAD

**Issue by SomnathS09** [aws-amplify/amplify-android#1419]
### Before opening, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-android/issues?q=is%3Aissue+) and [discussions](https://github.com/aws-amplify/amplify-android/discussions).

### Language and Async Model

Kotlin

### Amplify Categories

Storage

### Gradle script dependencies

<details>

```groovy
// Put output below this line
// Amplify core dependency
    implementation 'com.amplifyframework:core:1.18.0'

    implementation 'com.amplifyframework:aws-auth-cognito:1.18.0'
    implementation 'com.amplifyframework:aws-storage-s3:1.18.0'
```

</details>

### Environment information

<details>

```
# Put output below this line
Welcome to Gradle 6.5!

Here are the highlights of this release:
 - Experimental file-system watching
 - Improved version ordering
 - New samples

For more details see https://docs.gradle.org/6.5/release-notes.html


------------------------------------------------------------
Gradle 6.5
------------------------------------------------------------

Build time:   2020-06-02 20:46:21 UTC
Revision:     a27f41e4ae5e8a41ab9b19f8dd6d86d7b384dad4

Kotlin:       1.3.72
Groovy:       2.5.11
Ant:          Apache Ant(TM) version 1.10.7 compiled on September 1 2019
JVM:          1.8.0_242-release (JetBrains s.r.o 25.242-b01)
OS:           Windows 10 10.0 amd64

```

</details>

### Please include any relevant guides or documentation you're referencing

https://docs.amplify.aws/lib/storage/download/q/platform/android

### Describe the bug

There is no storage exception or error is thrown if Internet/Network is not available.
Amplify.Storage.downloadFile() just waits till network is available, and resumes the download process.
I need to tell the user if network is not available, not just postpone the task later.
Also it would be necessary to specify timeout for download request, currently there is no way to do it.

Amplify.Storage.listFiles() throws StorageException while Internet/Network is not available or lost, Need same functionality in downloadFile() or uploadFile().

Also let me know if there is any default request timeout in Storage.downloadFile(), I've waited for hours but no exception or error is thrown.

### Code Snippet

```java
// Put your code below this line.
Amplify.Storage.downloadFile(item.key, file, S3optionsForDownload,
                                { Log.i("StoryListViewModel", "Fraction completed: ${it.fractionCompleted}") },
                                { Log.i("StoryListViewModel", "Successfully downloaded: ${it.file.name}") },
                                { Log.e("StoryListViewModel", "Download Failure: ", it) }
                        )
```

### Log output

<details>

```
// Put your logs below this line


```

</details>

### GraphQL Schema

<details>

```graphql
// Put your schema below this line


```

</details>

**Comment by sdhuka** [comment]
@SomnathS09 listFile() is a synchronous api call hence it throws an exception incase of no network. UploadFile & DownloadFile api calls are queued and ran when network is available, these api calls are designed to gurantee completion in the future. There is no default timeout available, these api calls needs to be explicitly cancelled to remove from the queue.

**Comment by SomnathS09** [comment]
correct, but there is no direct access to queue using amplify-android sdk or how to cancel them?

**Comment by sdhuka** [comment]
You can use [cancel api on AWSS3StorageDownloadFileOperation](https://github.com/aws-amplify/amplify-android/blob/main/aws-storage-s3/src/main/java/com/amplifyframework/storage/s3/operation/AWSS3StorageDownloadFileOperation.java#L121) to cancel the request.

**Comment by SomnathS09** [comment]
Thanks @sdhuka , I'd tried it before
But for my use case : I'm having a list of files to download 
For this I firstly list files in bucket using listFiles() and then loop over each and fire up  

```
Amplify.Storage.downloadFile(item.key, file, S3optionsForDownload,
                                { Log.i("Download", "Fraction completed: ${it.fractionCompleted}") },
                                { Log.i("Download", "Successfully downloaded: ${it.file.name}") },
                                { Log.e("Download", "Download Failure: ", it) }
                        ) 
```
for each key.

I am maintaining **a list of AWSS3StorageDownloadFileOperation for all calls** made.
Now **in case of No network connectivity, I need to show the user that Internet in not available.**

so, I start cancelling each AWSS3StorageDownloadFileOperation in the list, but **when suddenly user gets connected to network, the AWSS3StorageDownloadFileOperation on which cancel() is yet to be called, starts downloading.**

**Comment by SomnathS09** [comment]
I think if somehow I'd access to [DownloadTransferListener](https://github.com/aws-amplify/amplify-android/blob/b53a5708f8068ad13e4b05086e02e856bd69d05e/aws-storage-s3/src/main/java/com/amplifyframework/storage/s3/operation/AWSS3StorageDownloadFileOperation.java#L166) ; I would have been able to call cancel() on WAITING_FOR_NETWORK event, right?

I do not want to extend it and create custom implementation, I this should be integrated in SDK.

Because, If user is not connected to Internet there is no sense in showing stuck progress bar we should show the appropriate message that network is not available.

**Comment by canberk-pensa** [comment]
@sdhuka 

I am also struggling with this issue, timeout/connection errors would be a great addition for my use case of uploadFile/downloadFile functions.

**Comment by abhayWo** [comment]
Is there any update on this issue?
I am also facing this issue in uploadFile(). When a connection is lost while uploading a file, the upload worker is canceled immediately and I can't get any kind of error that the operation was canceled.

so is there any issue while uploading the file if the internet is lost then singlePartUploadWorker is canceled without throwing any exception?

**Comment by souleateryx** [comment]
Same problem.There is no storage exception or error is thrown if Internet/Network is not available.
Amplify.Storage.uploadFile just waits till network is available, and resumes the download process.
